framework/src/vectorpostprocessors/WorkBalance.C

```diff
@@ -245,12 +255,24 @@ WorkBalance::execute()
 void
 WorkBalance::finalize()
 {
-  // Gather the results down to processor 0
-  _communicator.gather(0, static_cast<Real>(_local_num_elems), _num_elems);
-  _communicator.gather(0, static_cast<Real>(_local_num_nodes), _num_nodes);
-  _communicator.gather(0, static_cast<Real>(_local_num_dofs), _num_dofs);
-  _communicator.gather(0, static_cast<Real>(_local_num_partition_sides), _num_partition_sides);
-  _communicator.gather(0, _local_partition_surface_area, _partition_surface_area);
+  if (!_sync_to_all_procs)
+  {
+    // Gather the results down to processor 0
+    _communicator.gather(0, static_cast<Real>(_local_num_elems), _num_elems);
+    _communicator.gather(0, static_cast<Real>(_local_num_nodes), _num_nodes);
+    _communicator.gather(0, static_cast<Real>(_local_num_dofs), _num_dofs);
+    _communicator.gather(0, static_cast<Real>(_local_num_partition_sides), _num_partition_sides);
+    _communicator.gather(0, _local_partition_surface_area, _partition_surface_area);
+  }
+  else
+  {
+    // Gather the results down to all procs
+    _communicator.allgather(static_cast<Real>(_local_num_elems), _num_elems);
+    _communicator.allgather(static_cast<Real>(_local_num_nodes), _num_nodes);
+    _communicator.allgather(static_cast<Real>(_local_num_dofs), _num_dofs);
+    _communicator.allgather(static_cast<Real>(_local_num_partition_sides), _num_partition_sides);
+    _communicator.allgather(_local_partition_surface_area, _partition_surface_area);
```

Comment:
This really isn't the right idea here. You should be using`scatter` instead of `allgather`:
https://github.com/libMesh/libmesh/blob/master/include/parallel/parallel.h#L1240-L1247

Reply:
Nope.  `allgather()` is the right one.

All processors are contributing _one value_ that gets inserted in their `rank` position on every processor.  That's exactly what `allgather()` does.

Read up: http://mpitutorial.com/tutorials/mpi-scatter-gather-and-allgather/

Reply:
BTW - I thought of something else that could help out here. We have the `UserObject::spatialValue()` and associated AuxKernel that pulls those out. If we had a similar interface for VPPs, you could actually scatter _within_ your VPP and just store the scalar that you want to dump out. This would put the logic back into the VPP instead of the AuxKernel.

 I think we should consider fleshing this interface out more.

Reply:
👍 resolved